### PR TITLE
chore(gatsby-telemetry): Migrate package to TS

### PR DESCRIPTION
**Pull request by Eyas**
## Description

This PR migrates `packages/gatsby-telemetry/` to TypeScript.

## Related Issues

Part of #21995.

Note that this exposed #22531, which will be handled in a separate PR.

### CODE REVIEW

File: packages/gatsby-telemetry/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+const tick = (): void => {
+  flush()
+    .catch(console.error)
+    .then(() => setTimeout(tick, interval))
+}
+
+export const trackCli = (
+  input: string | undefined,
+  tags: {} | undefined,
```
Comment (LekoArts) [inline packages/gatsby-telemetry/src/index.ts]: `tags` and `opts` need to be optional, that's why the `lint` step is failing here: https://circleci.com/gh/gatsbyjs/gatsby/361624

**Comment by pieh** [comment]
I merged https://github.com/gatsbyjs/gatsby/pull/22613 (the `isTrackingEnabled` PR that changes `isTrackingEnabled` signature to be function returning boolean (instead of another function), so there will need to be updates in this PR

**Comment by Eyas** [comment]
Thanks. @pieh @LekoArts I just rebased and pushed. Can you take a look?

**Comment by Eyas** [comment]
FWIW this is updated now.

**Comment by Eyas** [comment]
`Record<string, string>` seems right for default tags (deocrateAll), though not all tags, e.g. tags can contain `{error: IErrorV1 | IErrorV1[], errorV2: IErrorV2 }`.

I'll leave tags as `unknown` since they seem be able to include arbitrary key/value pairs.

**Comment by LekoArts** [comment]
Thanks so much for the PR and sorry that we didn't get to merge it earlier! In the meantime a couple of other PRs changed the files bit by bit so that telemetry is actually done with the migration now!

**Pull request closed without merging**